Aqui empieza el curso de ML aplicado

In [5]:
import datetime
print(f'The last time is:  {datetime.datetime.now()}')

The last time is:  2026-04-06 15:52:49.459141


In [8]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import sklearn

Bueno, lo mas importante a la hora de estudiar ML y los modelos es entender los pasos de lo que estamos haciendo, entender la teoria, identificar el problema, tener los datos libros. 
Si tenemos un problema de clasificacion, un problema de prediccion o un problema de agrupacion. Los pasos que tenemos que considerar para el workflow de Scikit-learn son: 
1. Getting Data (find columns, labels, prepair train and test steps)
2. Choosing a model for our problem 
3. Ajustar el modelo a la data que tenemos y usarlo para obtener la predicción
4. Evaluar el modelo 
5. Experimentar para mejorar
6. Guardar el modelo para que alguien mas lo pueda usar, o yo mismo en el futuro. 

Para el primer ejemplo vamos a usar un dataset llamado heart-disease.csv

In [12]:
df = pd.read_csv("https://raw.githubusercontent.com/Juan-Vega18/ML_full_course/refs/heads/main/my_notebooks/data/heart-disease.csv?token=GHSAT0AAAAAADWIPU2OUAE5Y3CNWF7KIEE62OUHPWA")
df

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,57,0,0,140,241,0,1,123,1,0.2,1,0,3,0
299,45,1,3,110,264,0,1,132,0,1.2,1,0,3,0
300,68,1,0,144,193,1,1,141,0,3.4,1,2,3,0
301,57,1,0,130,131,0,1,115,1,1.2,1,1,3,0


In [20]:
#Ahora vamos a ver algunas propiedades de nuestro dataset
df.info()
df.shape
df.head(5)
#df['target'].unique()
#df['restecg'].unique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1


Encontramos un conjunto de 14 columnas con 303 filas. La fase de entender como y que especificamente compone nuestro dataset es absolutamente clave. Tenemos target que a pesar de que parece ser una variable numerica en realidad es booleana, el dataset esta completisimo, no tiene vacios etc. 

En el casode este proyecto, esa variable booleana es precisamente la que queremos predecir, si está o no enfermo del corazón, 1 significa si, 0 significa no. Las demas columnas se llamaran 'features' y las usaremos para intentar predecir a target. 

In [39]:
x = df.drop('target', axis=1) #Creamos los x de los parametros. eliminamos la columna target
y = df['target'] #Creamos la y que será nuestra predicción 
x.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2


In [ ]:
#Ahora vamos a revisar los valores de y que tenemos 
print(f'Los valores de la variable estan distribuidos de la siguiente forma: {y.value_counts()}')
y.unique()

Los valores de la variable estan distribuidos de la siguiente forma: target
1    165
0    138
Name: count, dtype: int64


array([1, 0])

Ahora, algo que es absolutamente importante en el ML es entender que la data debe ser divida en dos grupos, training set and test set. Es decir, un grupo de datos se usará para entrenar el modelo, el otro para probarlo. Usualmente se combina en proporciones como 80/20, 70/30, 75/25, tambien es posible tener el validation set que es un dataset extra que se usa durante el entrenamiento pero por ahora no lo vamos a usar. 

Scikit-learn ya incluye sklearn.model_selection.train_test_split que es la funcion de la libreria para dividir el dataset en training sets. 


In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.25)
print(X_train.shape)
print(type(X_train))
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

#Aqui ya dividimos la data, notemos que la funcion train_test_split asigna a 4 variables los 
#datos correspondientes y los subdivide en pandas.core.frame.DataFrame luego veremos que significa
# y que tiene de diferente con el pd.Dataframe tipico.

X_train

(227, 13)
<class 'pandas.core.frame.DataFrame'>
(76, 13)
(227,)
(76,)


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
122,41,0,2,112,268,0,0,172,1,0.0,2,0,2
196,46,1,2,150,231,0,1,147,0,3.6,1,0,2
290,61,1,0,148,203,0,1,161,0,0.0,2,1,3
279,61,1,0,138,166,0,0,125,1,3.6,1,1,2
81,45,1,1,128,308,0,0,170,0,0.0,2,0,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...
257,50,1,0,144,200,0,0,126,1,0.9,1,0,3
117,56,1,3,120,193,0,0,162,0,1.9,1,0,3
174,60,1,0,130,206,0,0,132,1,2.4,1,2,3
188,50,1,2,140,233,0,1,163,0,0.6,1,1,3


Ahora que tenemos los datos divididos es necesario escoger el modelo y los hiperparametros. Escoger el modelo depende normalmente de cual sea el problema que estamos trabajando, por ejemplo, Scikit-learn tiene recomendaciones si el problema es de clasificacion o predicción. 
En la web de scikit-learn tenemos un mapa muy util que explica que modelo usar, cuando y bajo que condiciones. 
La primera y mas importante es la cantidad de ejemplos (se recomienda mas de 50). 

Tenemos problemas de clasificacion, regresion, clustering y reduccion de la dimension. 
Aquí es donde el estudiante (yo) debería aprender bien y antes de que el curso acabe haber probado todos o la gran mayoría de los algoritmos, la programación no necesariamente es lo mas importante, lo que importa aquí es saber bien que modelo usar, cuando, y por qué. 

Tambien tenemos los hiperparametros del modelo, que si sabemos matematicamente lo que estamos haciendo, podemos tocarlos y revisar si el modelo mejora o no. Se van ajustando. Aquí tambien es muy importante que el estudiante entienda que está pasando en el interior de la maquina, de nada sirve tener el modelo, el codigo, resultados si no soy capaz de modificar parametros (no entiendo nada).

En el caso de este proyecto, estamos trabajando 